domain/user/state/state_test.go

```diff
@@ -4,81 +4,634 @@
+	c.Assert(err, jc.ErrorIs, usererrors.NotFound)
+}
+
+// TestGetUserByNameMultipleUsers asserts that we get a non-removed user when we try to
+// get a user by name that has multiple users with the same name.
+func (s *stateSuite) TestGetUserByNameMultipleUsers(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	// Add admin user.
+	adminUUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	adminUser := user.User{
+		Name:        "admin",
+		DisplayName: "admin",
+	}
+	err = st.AddUser(context.Background(), adminUUID, adminUser, adminUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Remove admin user.
+	err = st.RemoveUser(context.Background(), adminUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Add admin2 user.
+	admin2UUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	admin2User := user.User{
+		Name:        "admin",
+		DisplayName: "admin2",
+	}
+	err = st.AddUser(context.Background(), admin2UUID, admin2User, admin2UUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Get the user.
+	u, err := st.GetUserByName(context.Background(), "admin")
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(u.Name, gc.Equals, adminUser.Name)
+	c.Check(u.DisplayName, gc.Equals, admin2User.DisplayName)
+	c.Check(u.CreatorUUID, gc.Equals, admin2UUID)
+	c.Check(u.CreatedAt, gc.NotNil)
+}
+
+// TestGetUserByNameNotFound asserts that we get an error when we try to get a
+// user by name that does not exist.
+func (s *stateSuite) TestGetUserByNameNotFound(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	// Get the user.
+	_, err := st.GetUserByName(context.Background(), "admin")
+	c.Assert(err, jc.ErrorIs, usererrors.NotFound)
+}
+
+// TestRemoveUser asserts that we can remove a user from the database.
+func (s *stateSuite) TestRemoveUser(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	// Add admin user.
+	adminUUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	adminUser := user.User{
+		Name:        "admin",
+		DisplayName: "admin",
+	}
+	err = st.AddUser(context.Background(), adminUUID, adminUser, adminUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Add userToRemove.
+	userToRemoveUUID, err := user.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	userToRemove := user.User{
+		Name:        "userToRemove",
+		DisplayName: "userToRemove",
+	}
+	err = st.AddUser(context.Background(), userToRemoveUUID, userToRemove, adminUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Remove userToRemove.
+	err = st.RemoveUser(context.Background(), userToRemoveUUID)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Check that the user was removed correctly.
+	db := s.DB()
+
+	row := db.QueryRow(`
+SELECT removed
+FROM user
+WHERE uuid = ?
+	`, userToRemoveUUID)
+	c.Assert(row.Err(), jc.ErrorIsNil)
+
+	var removed bool
+	err = row.Scan(&removed)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(removed, gc.Equals, true)
+}
+
+// TestSetPasswordHash asserts that we can set a password hash for a user.
+func (s *stateSuite) TestSetPasswordHash(c *gc.C) {
```

Comment:
Should we add a test for setting a password hash on a user that has had their auth disabled?

Reply:
Hmmm,  looks like we don't have function that disabled auth.